crates/rattler_repodata_gateway/src/gateway/sharded_subdir/index.rs

```diff
@@ -0,0 +1,352 @@
+use reqwest_middleware::ClientWithMiddleware;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use std::{io::Write, path::Path, str::FromStr, time::SystemTime};
+use tempfile::NamedTempFile;
+use tokio::fs::File;
+use tokio::io::{AsyncReadExt, BufReader};
+use url::Url;
+
+/// Magic number that identifies the cache file format.
+const MAGIC_NUMBER: &[u8] = b"SHARD-CACHE-V1";
+
+const REPODATA_SHARDS_FILENAME: &str = "repodata_shards.msgpack.zst";
+
+// Fetches the shard index from the url or read it from the cache.
+pub async fn fetch_index(
+    client: ClientWithMiddleware,
+    channel_base_url: &Url,
+    token_client: &TokenClient,
+    cache_dir: &Path,
+    concurrent_requests_semaphore: Arc<tokio::sync::Semaphore>,
+    reporter: Option<&dyn Reporter>,
+) -> Result<ShardedRepodata, GatewayError> {
+    async fn from_response(
+        cache_path: &Path,
+        policy: CachePolicy,
+        response: Response,
+        reporter: Option<(&dyn Reporter, usize)>,
+    ) -> Result<ShardedRepodata, GatewayError> {
+        // Read the bytes of the response
+        let response_url = response.url().clone();
+        let bytes = response.bytes_with_progress(reporter).await?;
+
+        if let Some((reporter, index)) = reporter {
+            reporter.on_download_complete(&response_url, index);
+        }
+
+        // Decompress the bytes
+        let decoded_bytes = Bytes::from(super::decode_zst_bytes_async(bytes).await?);
+
+        // Write the cache to disk if the policy allows it.
+        let cache_fut = if policy.is_storable() {
+            write_shard_index_cache(cache_path, policy, decoded_bytes.clone())
+                .map_ok(Some)
+                .map_err(|e| {
+                    GatewayError::IoError(
+                        format!(
+                            "failed to create temporary file to cache shard index to {}",
+                            cache_path.display()
+                        ),
+                        e,
+                    )
+                })
+                .left_future()
+        } else {
+            // Otherwise delete the file
+            tokio::fs::remove_file(cache_path)
+                .map_ok_or_else(
+                    |e| {
+                        if e.kind() == std::io::ErrorKind::NotFound {
+                            Ok(None)
+                        } else {
+                            Err(GatewayError::IoError(
+                                format!(
+                                    "failed to remove cached shard index at {}",
+                                    cache_path.display()
+                                ),
+                                e,
+                            ))
+                        }
+                    },
+                    |_| Ok(None),
+                )
+                .right_future()
+        };
+
+        // Parse the bytes
+        let parse_fut = tokio_rayon::spawn(move || rmp_serde::from_slice(&decoded_bytes))
```

Comment:
I have read a little about rayon vs tokio - is it worth it? From what I read it seems most useful in CPU intense workloads while we are presumably mostly bound by IO.

https://www.reddit.com/r/rust/comments/xec77k/rayon_or_tokio_for_heavy_filesystem_io_workloads/

Reply:
Removed rayon, seems fine!